File: src/ingest/web.ts
import { config } from '../config.js';
import { logger } from '../logger.js';
import type { ConvertedDoc } from './index.js';

interface DoclingSourceResponse {
  document?: { md_content?: string; filename?: string };
  status?: string;
  errors?: Array<{ error_message: string }>;
}

/**
 * URL を Docling Serve に渡し Markdown を取得する。
 * - HTTP / HTTPS のみ許可
 * - HTML / PDF / 画像のいずれでも 1 経路で扱える
 * - OCR は HTML では発火しないが、PDF / 画像 URL では自動で発火
 */
export async function convertWeb(url: string): Promise<ConvertedDoc> {
  if (!/^https?:\/\//i.test(url)) throw new Error(`Not an HTTP(S) URL: ${url}`);

  const body = {
    sources: [{ kind: 'http', url }],
    options: {
      from_formats: ['pdf', 'image', 'docx', 'pptx', 'html', 'md'],
      to_formats: ['md'],
      do_ocr: true,
      ocr_engine: 'easyocr',
      ocr_lang: ['ja', 'en'],
      table_mode: 'accurate',
      image_export_mode: 'placeholder',
      abort_on_error: false,
      return_as_file: false,
    },
  };

  logger.info({ url }, 'Calling Docling Serve (URL source)');
  const res = await fetch(`${config.DOCLING_URL}/v1alpha/convert/source`, {
    method: 'POST',
    headers: { 'content-type': 'application/json' },
    body: JSON.stringify(body),
    signal: AbortSignal.timeout(10 * 60 * 1000),
  });
  if (!res.ok) throw new Error(`Docling URL convert failed (${res.status}): ${await res.text()}`);

  const json = await res.json() as DoclingSourceResponse;
  if (json.status !== 'success' && json.errors?.length) {
    throw new Error(`Docling errors: ${json.errors.map(e => e.error_message).join('; ')}`);
  }
  const md = json.document?.md_content ?? '';
  if (!md) throw new Error(`Empty markdown from URL: ${url}`);

  return {
    source: url,
    markdown: md,
    metadata: {
      kind: 'web',
      url,
      filename: json.document?.filename,
      fetchedAt: new Date().toISOString(),
    },
  };
}

/**
 * `.urls` 拡張子のテキストファイル（行ごと 1 URL、`#` で始まる行はコメント）を読み URL 配列を返す。
 * `expandPath` で展開され、各 URL は `convertWeb` で個別取込される。
 */
export async function readUrlList(filePath: string): Promise<string[]> {
  const { readFile } = await import('node:fs/promises');
  const raw = await readFile(filePath, 'utf-8');
  return raw.split(/\r?\n/).map(l => l.trim()).filter(l => l && !l.startsWith('#'));
}

File: crates/ingest/src/web.rs
use super::{ConvertedDoc, DocMetadata};
use rag_common::{AppError, Config, Result};
use serde::Deserialize;
use serde_json::json;
use std::path::Path;
use std::time::Duration;

#[derive(Debug, Deserialize)]
struct DoclingSourceResponse {
    #[serde(default)]
    document: Option<DoclingDoc>,
    #[serde(default)]
    status: Option<String>,
    #[serde(default)]
    errors: Option<Vec<DoclingError>>,
}

#[derive(Debug, Deserialize)]
struct DoclingDoc {
    #[serde(default)]
    md_content: Option<String>,
}

#[derive(Debug, Deserialize)]
struct DoclingError {
    #[serde(default)]
    error_message: String,
}

pub async fn convert_web(url: &str) -> Result<ConvertedDoc> {
    if !crate::is_url(url) {
        return Err(AppError::Validation(format!("Not an HTTP(S) URL: {url}")));
    }
    let cfg = Config::get();
    let body = json!({
        "sources": [{ "kind": "http", "url": url }],
        "options": {
            "from_formats": ["pdf", "image", "docx", "pptx", "html", "md"],
            "to_formats": ["md"],
            "do_ocr": true,
            "ocr_engine": "easyocr",
            "ocr_lang": ["ja", "en"],
            "table_mode": "accurate",
            "image_export_mode": "placeholder",
            "abort_on_error": false,
            "return_as_file": false,
        },
    });

    let endpoint = format!(
        "{}/v1alpha/convert/source",
        cfg.docling_url.trim_end_matches('/')
    );
    let client = reqwest::Client::builder()
        .timeout(Duration::from_secs(600))
        .build()
        .map_err(AppError::http)?;

    let resp = client.post(&endpoint).json(&body).send().await?;
    let status = resp.status();
    if !status.is_success() {
        let text = resp.text().await.unwrap_or_default();
        return Err(AppError::Docling(format!("status={status}, body={text}")));
    }
    let parsed: DoclingSourceResponse = resp.json().await?;
    if parsed.status.as_deref() != Some("success") {
        if let Some(errs) = parsed.errors {
            let msg = errs
                .iter()
                .map(|e| e.error_message.clone())
                .collect::<Vec<_>>()
                .join("; ");
            return Err(AppError::Docling(msg));
        }
    }
    let md = parsed
        .document
        .and_then(|d| d.md_content)
        .unwrap_or_default();
    if md.is_empty() {
        return Err(AppError::Docling(format!("Empty markdown from URL: {url}")));
    }

    Ok(ConvertedDoc {
        source: url.to_string(),
        markdown: md,
        metadata: DocMetadata::Web {
            url: url.to_string(),
        },
    })
}

pub async fn read_url_list(path: &Path) -> Result<Vec<String>> {
    let bytes = tokio::fs::read(path).await?;
    let raw = String::from_utf8(bytes).map_err(AppError::parse)?;
    let mut out = Vec::new();
    for line in raw.split('\n') {
        let t = line.trim().trim_end_matches('\r').trim();
        if t.is_empty() || t.starts_with('#') {
            continue;
        }
        out.push(t.to_string());
    }
    Ok(out)
}
